cpp/include/cuspatial/geometry/vec_3d.hpp

```diff
@@ -0,0 +1,184 @@
+   */
+  friend vec_3d<T> CUSPATIAL_HOST_DEVICE operator+(vec_3d<T> const& a, vec_3d<T> const& b)
+  {
+    return vec_3d<T>{a.x + b.x, a.y + b.y, a.z + b.z};
+  }
+
+  /**
+   * @brief Element-wise subtraction of two 3D vectors.
+   */
+  friend vec_3d<T> CUSPATIAL_HOST_DEVICE operator-(vec_3d<T> const& a, vec_3d<T> const& b)
+  {
+    return vec_3d<T>{a.x - b.x, a.y - b.y, a.z - b.z};
+  }
+
+  /**
+   * @brief Invert a 3D vector.
+   */
+  friend vec_3d<T> CUSPATIAL_HOST_DEVICE operator-(vec_3d<T> const& a)
+  {
+    return vec_3d<T>{-a.x, -a.y, -a.z};
+  }
+
+  /**
+   * @brief Scale a 3D vector by a factor @p r.
+   */
+  friend vec_3d<T> CUSPATIAL_HOST_DEVICE operator*(vec_3d<T> vec, T const& r)
+  {
+    return vec_3d<T>{vec.x * r, vec.y * r, vec.z * r};
+  }
+
+  /**
+   * @brief Scale a 3d vector by ratio @p r.
+   */
+  friend vec_3d<T> CUSPATIAL_HOST_DEVICE operator*(T const& r, vec_3d<T> vec) { return vec * r; }
+
+  /**
+   * @brief Translate a 3D point
+   */
+  friend vec_3d<T>& CUSPATIAL_HOST_DEVICE operator+=(vec_3d<T>& a, vec_3d<T> const& b)
+  {
+    a.x += b.x;
+    a.y += b.y;
+    a.z += b.z;
+    return a;
+  }
+
+  /**
+   * @brief Translate a 3D point
+   */
+  friend vec_3d<T>& CUSPATIAL_HOST_DEVICE operator-=(vec_3d<T>& a, vec_3d<T> const& b)
+  {
+    return a += -b;
+  }
+
+  /**
+   * @brief Less than operator for two 3D points.
+   *
+   * Orders two points first by x, then by y.
+   */
+  friend bool CUSPATIAL_HOST_DEVICE operator<(vec_3d<T> const& lhs, vec_3d<T> const& rhs)
+  {
+    if (lhs.x < rhs.x) return true;
+    if (lhs.x > rhs.x) return false;
+    if (lhs.y < rhs.y) return true;
+    if (lhs.y > rhs.y) return false;
+    return lhs.z < rhs.z;
+  }
+
+  /**
+   * @brief Greater than operator for two 3D points.
+   */
+  friend bool CUSPATIAL_HOST_DEVICE operator>(vec_3d<T> const& lhs, vec_3d<T> const& rhs)
+  {
+    return rhs < lhs;
+  }
+
+  /**
+   * @brief Less than or equal to operator for two 3D points.
+   */
+  friend bool CUSPATIAL_HOST_DEVICE operator<=(vec_3d<T> const& lhs, vec_3d<T> const& rhs)
+  {
+    return !(lhs > rhs);
+  }
+
+  /**
+   * @brief Greater than or equal to operator for two 3D points.
+   */
+  friend bool CUSPATIAL_HOST_DEVICE operator>=(vec_3d<T> const& lhs, vec_3d<T> const& rhs)
+  {
+    return !(lhs < rhs);
+  }
+};
+
+// Deduction guide enables CTAD
+template <typename T>
+vec_3d(T x, T y, T z) -> vec_3d<T>;
+
+/**
+ * @brief Compute dot product of two 3D vectors.
+ */
+template <typename T>
+T CUSPATIAL_HOST_DEVICE dot(vec_3d<T> const& a, vec_3d<T> const& b)
+{
+  return a.x * b.x + a.y * b.y + a.z * b.z;
+}
+
+/**
+ * @brief Compute 3D determinant of a 2x2 matrix with column vectors @p a and @p b.
+ */
+template <typename T>
+vec_3d<T> CUSPATIAL_HOST_DEVICE cross_product(vec_3d<T> const& a, vec_3d<T> const& b)
+{
+  return {a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x};
+}
+
+}  // namespace cuspatial
```

Comment:
You must be failing clang-format checks if your files are not ending correctly. Make sure you either run `pre-commit run` before committing, or just enable "format on save" for C++ files in your editor.

```suggestion
}  // namespace cuspatial

```